### DISCUSSION TITLE
Persist middleware: can I await on a set operation?

### DISCUSSION THREAD

**Discussion by vaharoni** [pmndrs/zustand#2596] [Q&A]
I'm using `AsyncStorage` and would like to use the Persist middleware. I have a React page where I'd like some user event to save the data to the store and navigate to another page when this completes. The [guide](https://github.com/pmndrs/zustand/blob/main/docs/integrations/persisting-store-data.md#how-can-i-check-if-my-store-has-been-hydrated) discusses the recipe of using `onRehydrateStorage` to set a flag when the hydration process finishes. But I did not see a reference for awaiting for the set operation to complete. In [this discussion](https://github.com/pmndrs/zustand/discussions/287), the OP calls `await set({ fish: get().fish + 1 })` inside the `persist` callback, but typescript does not agree that the persist's `set` function returns a promise, so I'm assuming such usage is incorrect.

What's the best practice here? Does the API expose a way to await the set operation? Or do I need to implement a store action that calls `AsyncStorage.setItem`, using the middleware mostly for reading but not for setting? Or is the idea perhaps to assume it will eventually be consistent?

**Comment by dai-shi** [comment]
Can anyone help?

**Comment by dbritto-dev** [comment]
@vaharoni I believe that you should create a custom storage

**Reply by vaharoni** [reply]
Thank you @dbritto-dev. Would it help, though? My understanding is that custom storage is an object that satisfies the `getItem`, `setItem`, and `removeItem` interface. These indeed can be async, but isn't the Persist middleware still the one that decides when to invoke them per its internal lifecycle, and I'm still supposed to use the store's hook as my API for state changes? 

If I have a store like this:
```typescript
export const useDisplayNameStore = create<DisplayNameState>()(
  persist(
    (set) => ({
      _hasHydrated: false,
      displayName: '',
      actions: {
        setHasHydrated: _hasHydrated => set({ _hasHydrated }),
        setDisplayName: displayName => set({ displayName })
      }
    }),
    {
      name: 'display-name',
      storage: createJSONStorage(() => AsyncStorage),
      onRehydrateStorage: state => state.actions.setHasHydrated(true)
    },
  ),
)
```

My hope was that in a component I can simply use my store actions:
```typescript
export default function Onboarding() {
  const router = useRouter();
  const { setDisplayName } = useDisplayNameStore(state => state.actions);

  async function onUpdate() {
    setDisplayName('test')
    // I want to await here until the store persists itself and the storage is in sync with the state
    router.navigate({ pathname: 'somewhere' })
  }

  return (
    // ...
  )
}
```

and that the Persist middleware will take care of the rest. But I'm not sure how to tell it "persist the store now based on the state changes" and wait for this process to complete.

**Reply by dbritto-dev** [reply]
@vaharoni btw, is react native app or nextjs app?

**Reply by dbritto-dev** [reply]
@vaharoni @dai-shi seems that setItem is not supporting async operations but getItem does

**Reply by vaharoni** [reply]
@dbritto-dev It's a React Native app using Expo

**Reply by vaharoni** [reply]
@dbritto-dev in the "[How can I use a custom storage engine](https://github.com/pmndrs/zustand/blob/main/docs/integrations/persisting-store-data.md#how-can-i-use-a-custom-storage-engine)" section of the FAQs it suggests `setItem` of a custom storage engine can be async. Though I'm not sure custom storage engine solves the key issue here of the store lifecycle per my previous comment.

**Reply by dbritto-dev** [reply]
@vaharoni yeah, but in the code doesn't do that but. The types tell you something else, so I believe we should fix that

**Reply by dbritto-dev** [reply]
@vaharoni it should fix that

**Comment by vaharoni** [comment]
Eventually I noticed `persist` behaved funky for me also in the hydration phase - it seemed to ignore for no obvious reason some values that I verified were present in the `AsyncStore`. This totally could have been a bug on my end, but I ended up moving away from `persist` and implementing a code with less magic. That said, I saved a branch on the side with `persist` to help further test and investigate the issue if necessary.

Here's the code I ended up with. It's likely only suitable for small stores, as saving the state independently for each action could get old very quickly.
```typescript
import { createStore, useStore } from 'zustand'
import AsyncStorage from '@react-native-async-storage/async-storage'

interface DisplayNameState {
  hydrated: boolean
  displayName: string
  actions: {
    setDisplayName: (displayName: string) => Promise<void>
  }
}

const STORAGE_ITEM = 'display-name-store';

const displayNameStore = createStore<DisplayNameState>()(
  (set, get) => ({
    hydrated: false,
    displayName: '',
    actions: {
      // This being async allows us to await
      setDisplayName: async (displayName) => { 
        set({ displayName });
        // removing actions - they are not serialized and we don't want them to override our functions during hydration
        const { actions, ...newState } = get();
        await AsyncStorage.setItem(STORAGE_ITEM, JSON.stringify(newState));
      }
    }
  })
)

export const useDisplayNameStore = (selector: (selector: DisplayNameState) => any) => {
  const { getState, setState } = displayNameStore;
  if (!getState().hydrated) {
    AsyncStorage.getItem(STORAGE_ITEM)
      .then(data => {
        const object = data ? JSON.parse(data) : {}
        const initialState = { ...object, hydrated: true }
        setState(initialState)
      })
  }
  return useStore(displayNameStore, selector)
}
```
Usage:
```typescript
export default function MyComponent() {
  const { setDisplayName } = useDisplayNameStore(state => state.actions)
  const router = useRouter();

  // ...

  async function onSomeEvent() {
    // await here guarantees it is saved to the store before navigating
    await setDisplayName('test')
    router.navigate('somewhere')
  }

  // ...
}
```

**Reply by dbritto-dev** [reply]
@vaharoni to avoid function being serialized you should use `partialize`

**Reply by vaharoni** [reply]
Thanks @dbritto-dev. `partialize` is available only in the `persist` middleware, I believe, correct? I don't think this was what had caused my issues with the middleware (I gave `partialize` a try when I wrongly thought the `_hasHydrated` state should not persist). But I can play with it again and see if I missed something.

**Reply by dbritto-dev** [reply]
@vaharoni btw, did you check it out the example that I shared you?

**Reply by vaharoni** [reply]
Which example @dbritto-dev? I don't see one in the thread.

**Reply by dbritto-dev** [reply]
@vaharoni here you go -> https://stackblitz.com/edit/nextjs-and-zustand-persist?file=README.md

**Reply by vaharoni** [reply]
Thank you for sending, @dbritto-dev. This uses `LocalStorage` in a web context on next.js. `LocalStorage` is a synchronous storage engine, so you don't need to incorporate async-await considerations. As I am using `AsyncStorage` on React Native, I don't think this applies here. I believe the asynchronous nature of that storage engine is where all the issues stem from.

**Reply by dbritto-dev** [reply]
@vaharoni actually, you can use the same approach

**Reply by melutovich** [reply]
@vaharoni @dbritto-dev @dai-shi 

The way I understand this discussion (and the code) is that at least for now zustand for asynchronous storage does not support neither the ability to await set nor to register a listener to be informed when the storage has completed (or failed), correct?
(Usage discussed here and the originally referenced discussion to ensure storage is saved before finalizing a logout or before navigating)

The documentation for custom engine has
```
setItem: async (name: string, value: string): Promise<void> => {
    console.log(name, 'with value', value, 'has been saved')
    await set(name, value)
  },
```

As the returned Promise gets ignored should the documentation indicate the Promise is ignored?
Ignored (if I understand the code correctly) like in: https://github.com/pmndrs/zustand/blob/0a4f9d0f71477c5ef399191acc19e25674d0d3c4/src/middleware/persist.ts#L218

**Reply by dbritto-dev** [reply]
@melutovich persist does support asynchronous storage but does not perform any checks, in order to implement some checks you need to use a custom storage

```ts
export interface StateStorage {
  getItem: (name: string) => string | null | Promise<string | null>
  setItem: (name: string, value: string) => unknown | Promise<unknown>
  removeItem: (name: string) => unknown | Promise<unknown>
}
```

**Reply by melutovich** [reply]
@dbritto-dev Understood and I'll have to go that path for now, however that does mean that a consumer of the store must now work together with this custom storage, not very ideal.

**Reply by dbritto-dev** [reply]
@melutovich that goes to user land not to library land, zustand is unopinionated

**Reply by melutovich** [reply]
@dbritto-dev
zustand swallowing the Promise in 'void setItem()' (and presumably other locations) I'll argue (hopefully for clarification) does mean that zustand is opinionated, it is saying that 'you can use a custom storage and one that is an asynchronous storage but be aware that any Promise from any async setItem gets swallowed.  Want to switch from one custom storage to another that is also async just remember to deal in each one appropriately with the shallowed Promise.'

I think in those locations it should be having 'return setItem()' which was discussed in the much earlier discussion of https://github.com/pmndrs/zustand/discussions/287

The old code had just 'setItem()' as referenced in that older discussion, I'm not sure why it was switched to 'void setItem()' instead of 'return setItem()'

**Reply by dbritto-dev** [reply]
@melutovich because that's an expected behavior storage like `localStorage`, `sessionStorage` and `AsyncStorage` does not return something, so returning values in `setItem` is not correct
